File: sol/solution/0001-0100/0008. String to Integer (atoi)/Solution.py
class Solution:
    def myAtoi(self, s: str) -> int:
        i, n, sign, result = 0, len(s), 1, 0
        while i < n and s[i] == ' ': i += 1
        if i < n and s[i] in '+-': sign = -1 if s[i] == '-' else 1; i += 1
        while i < n and s[i].isdigit():
            digit = int(s[i])
            if result > (2**31 - 1 - digit) // 10:
                return -2**31 if sign == -1 else 2**31 - 1
            result = result * 10 + digit
            i += 1
        return sign * result

File: sol/solution/0001-0100/0008. String to Integer (atoi)/Solution.kt
class Solution {
    fun myAtoi(s: String): Int {
        var i = 0
        val n = s.length
        var sign = 1
        var result = 0
        while (i < n && s[i] == ' ') i++
        if (i < n && (s[i] == '+' || s[i] == '-')) {
            sign = if (s[i] == '-') -1 else 1
            i++
        }
        while (i < n && s[i].isDigit()) {
            val digit = s[i] - '0'
            if (result > (Int.MAX_VALUE - digit) / 10)
                return if (sign == 1) Int.MAX_VALUE else Int.MIN_VALUE
            result = result * 10 + digit
            i++
        }
        return result * sign
    }
}